openedx/core/djangoapps/agreements/models.py

```diff
@@ -22,3 +22,15 @@ class IntegritySignature(TimeStampedModel):
     class Meta:
         app_label = 'agreements'
         unique_together = ('user', 'course_key')
+
+
+class LTIPIISignature(models.Model):
+    """
+    This model stores a user's acknowledgement to share PII via LTI tools in a particular course.
+    """
+    course_key = CourseKeyField(max_length=255, db_index=True)
+    user_id = models.IntegerField()
+    lti_tools_hash = models.IntegerField()
```

Comment:
Could you leave a comment about what the purpose of this field is, please? I think it could be somewhat confusing why we're storing a hash, so a comment will help future readers of this code.

Maybe something like the following, or whatever you prefer.

> We are storing a hash of the LTI tools that receive PII that a learner has acknowledged. We do this so that we can quickly and easily compare this hash with the `lti_tools_hash` of the `LTIPIITool` model. This allows us to reduce response time and decrease any impact on unit rendering time.

Reply:
I created the lti_tools_hash field because I followed the ticket description directly. Instead, I think storing the list of LTI Tools would make more sense since a course's set of tools won't constantly change. Does this align with what you had in mind?

Reply:
Ah, I didn't meant to imply that I don't think the field is necessary. I think it is, but I wanted to call out that it might not be obvious for everyone what the purpose of it is, so I think it's valuable to explain why we're adding it and how we plan to use it. That's why I recommended we leave a comment.

I agree that storing the list of LTI tools somehow make sense. After all, we need to store the LTI tools that a learner "acknowledged" would receive their PII when they sign the acknowledgment that's displayed to them in the courseware. And ordinarily, I think that would be enough. 

The issue is that we need an answer to the question, "Should we show the learner the LTI PII acknowledgment dialog" every single time they render the courseware. That's a lot! Let's imagine that we are storing `lti_tools` on this model - `LTIPIISignature` - too. The way that we would answer that question is by checking the `lti_tools` field of the `LTIPIITools` model against the `lti_tools` field of this model - `LTIPIISignature`. If they're equal, there's no need to display the dialog. If they're not, then we need to redisplay the dialog because the list of PII receiving LTI tools in the course has changed.

It's really important that the courseware is performant and quick. Therefore, the reason we're also storing a hash of that list is to speed up that comparison. Because the hashes are integers, comparing them will be faster than comparing two lists. If we store the hash in this model, we don't really have a need for the list of LTI tools, which is why it's omitted in the ticket.

On the other hand, storing a list in addition to the hash in this model could make sense because it would make debugging issues in the future easier. What do you think?

Reply:
That's true, but it would depend how we're using it. If we ever want to actually know what LTI tools a learner acknowledged, then a hash will not give us that information. We can't "unhash" that value to get a human-readable list of LTI tools. In that way, I'd say they're not redundant. I think it depends on their function.

I don't think so. They're independent. `LTIPIITool` represents what LTI tools are currently active in the course that are configured to receive PII from edX. That list has nothing to do with the user. `lti_tools` is a list of LTI tool URLs for those LTI tools receiving PII from edX. `lti_tools_hash` is a hash of that list.

`LTIPIISignature`, on the other hand, represents an acknowledgment that a learner made that a certain set of LTI tools will receive their PII in a particular course from edX. The `lti_tools_hash` is a hash of that list. I would say, most of the time` `LTIPIITool.lti_tools_hash` and `LTIPIISignature.lti_tools_hash` will be equal after a learner signs and acknowledgment. But let's say I log out for a week. In that week, a course team member adds a new tool. Now, `LTIPIITool.lti_tools_hash` and `LTIPIISignature.lti_tools_hash` are no longer equal, because I originally signed an acknowledgment that did not include that new tool. `LTIPIITool.lti_tools_hash` will be a new value that represents the new list with the additional new tool in it. When I log back into the courseware, I'll be asked to acknowledge again, because the hashes are unequal. At that point, we'd update the `LTIPIISignature.lti_tools_hash` field. Does that make sense? Or am I missing something?